src/main/kotlin/io/sip3/salto/ce/sdp/SdpHandler.kt

```diff
@@ -0,0 +1,175 @@
+
+            if (isFilled()) {
+                sendSdpSession(this)
+            }
+        }
+    }
+
+    private fun sendSdpSession(sdpSessionDescription: SdpSessionDescription) {
+        val requestMediaDescription = sdpSessionDescription.request.getMediaDescription("audio")
+        val responseMediaDescription = sdpSessionDescription.response.getMediaDescription("audio")
+
+        val callId = sdpSessionDescription.callId
+        val codec = getCodec(requestMediaDescription, responseMediaDescription)
+        val requestSdpSession = createSdpSession(callId, requestMediaDescription, codec)
+        val responseSdpSession = createSdpSession(callId, responseMediaDescription, codec)
+
+        // Send to Management Socket
+        vertx.eventBus().send(RoutesCE.sdp_session, listOf(requestSdpSession, responseSdpSession), USE_LOCAL_CODEC)
+    }
+
+    private fun getCodec(request: MediaDescriptionField, response: MediaDescriptionField): Codec {
+        val payloadTypes = request.payloadTypes.intersect(response.payloadTypes.asIterable())
+        val payloadType = payloadTypes.first()
+        val payload: RtpMapAttribute? = response.getFormat(payloadType)
+                ?: request.getFormat(payloadType)
+
+        val codec = codecs[payload?.codec]
+                ?: codecs.values.firstOrNull { it.payloadType.toInt() == payloadType }
+                ?: Codec()
+
+        // Replace with values from SDP
+        codec.payloadType = payloadType.toByte()
+        payload?.let { codec.clockRate = it.clockRate }
+        return codec
+    }
+
+    private fun createSdpSession(callId: String, mediaDescription: MediaDescriptionField, codec: Codec): SdpSession {
+        return SdpSession().apply {
+            id = getSdpSessionId(mediaDescription)
+            timestamp = System.currentTimeMillis()
+
+            payloadType = codec.payloadType
+            codecName = codec.name
+            clockRate = codec.clockRate
+            codecIe = codec.ie
+            codecBpl = codec.bpl
+
+            this.callId = callId
+        }
+    }
+
+    private fun getSdpSessionId(mediaDescription: MediaDescriptionField): Long {
+        val addr = IpUtil.convertToInt(mediaDescription.connection.address).toLong()
+        val port = mediaDescription.port.toLong()
+        return (addr shl 32) or port
+    }
+
+    private fun getSessionDescription(message: SIPMessage): SessionDescription? {
+        if (message.contentTypeHeader?.mediaSubType == "sdp") {
+            return SessionDescriptionParser.parse(message.messageContent)
+        } else {
+            message.multipartMimeContent?.contents?.forEach { mimeContent ->
+                if (mimeContent.matches("sdp")) {
+                    return SessionDescriptionParser.parse(mimeContent.content.toString())
+                }
+            }
+        }
+
+        return null
+    }
+
+    private class SdpSessionDescription(val callId: String) {
+
+        lateinit var request: SessionDescription
```

Comment:
there is no sense to make these two `lateinit`. `::request.isInitialized` is ugly and slow. Given that we won't read this properties - we can just make it nullable.